### ISSUE TITLE
Host ports are not mapped when kubenet is enabled in e2e-node-test

### ISSUE THREAD

**Issue by coufon** [kubernetes/kubernetes#29039]
In e2e-node-test we want to use a standalone Cadvisor pod to monitoring kubelet/docker-daemon resource usage. We create the Cadvisor pod in the e2e-node-test framework and have 8090(host):8080(container) port mapping. However, the pod runs successfully but the port is not mapped.

We tried disabling 'kubenet' (--disable-kubenet=true) and found this problem was gone. We tried on three test images (test-e2e-node-coreos-alpha-1068-20160707-image, test-e2e-node-containervm-v20160321-image, test-e2e-node-ubuntu-trusty-docker9-v1-image) and the results were the same.

When kubenet is enabled, port mapping is normal when creating the pod by docker directly.

And for the image 'test-e2e-node-containervm-v20160321-image', we can not access the Cadvisor endpoint in container via HTTP, although we can access it inside the container.

@Random-Liu

**Comment by Random-Liu** [comment]
I've debugged this with @coufon just now. We found that `hostport` would never work in node e2e test when `kubenet` is enabled. And once we disable `kubenet`, everything works fine after that.

We are not sure whether this is an issue in kubenet or node e2e test image. Maybe because we didn't set the docker flag `--iptables=false --ip-masq=false`(https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L308) in the node e2e test.

However, after we manually update the docker flag and restart docker in the existing instance, it still doesn't work. Maybe because something has already been massed up during the first run.

/cc @kubernetes/sig-network @kubernetes/sig-node

**Comment by Random-Liu** [comment]
BTW, if `--iptables=false --ip-masq=false` is a must have, we may want to check it in `kubelet verify`.

XRef https://github.com/kubernetes/kubernetes/issues/28408.

**Comment by bprashanth** [comment]
hmm can you look at your iptables rules for the kube-hostports chain? something like 
`sudo iptables -t nat -L | grep -i KUBE-HP -C 10`
should be enough

what's it pointing to? can you curl the endpoint ip etc

**Comment by yujuhong** [comment]
I created a simple http server pod with container port 8080 and host port 10234

```console
$ sudo iptables -t nat -L | grep -i KUBE-HP -C 10
target     prot opt source               destination         
MASQUERADE  all  --  172.17.0.0/16        anywhere            
MASQUERADE  all  --  anywhere            !10.0.0.0/8           /* kubenet: SNAT for outbound traffic from cluster */ ADDRTYPE match dst-type !LOCAL
MASQUERADE  all  --  loopback/8           anywhere             /* SNAT for localhost access to hostports */

Chain DOCKER (2 references)
target     prot opt source               destination         

Chain KUBE-HOSTPORTS (2 references)
target     prot opt source               destination         
KUBE-HP-63IXABCL5DSDLSFA  tcp  --  anywhere             anywhere             /* ok-test-e2e-node-containervm-v20160321-image_default hostport 10234 */ tcp dpt:10234

Chain KUBE-HP-63IXABCL5DSDLSFA (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.180.0.43          anywhere             /* ok-test-e2e-node-containervm-v20160321-image_default hostport 10234 */
DNAT       tcp  --  anywhere             anywhere             /* ok-test-e2e-node-containervm-v20160321-image_default hostport 10234 */ tcp to:10.180.0.43:8080

Chain KUBE-MARK-MASQ (1 references)
target     prot opt source               destination         
```

Curling the pod_IP:container_port works.
```console
$ curl http://10.180.0.43:8080
ok
```
Curling the localhost:host_port failed hangs.
```console
$ curl http://localhost:10234
```

**Comment by Random-Liu** [comment]
I've tried with `--iptables=false --ip-masq=false`, and it still doesn't work.
I started a pod using hostport 8001, I didn't work when kubenet is enabled:
```
$ curl localhost:8001
```
My docker configuration:
```
$ ps aux | grep docker
root      2215  2.8  1.2 684160 46956 ?        Sl   02:41   0:06 /usr/bin/docker daemon -p /var/run/docker.pid -H u
nix:// -r=false --iptables=false --ip-masq=false
```
Iptable rules:
```
$ sudo iptables -t nat -L | grep -i KUBE-HP -C 10
target     prot opt source               destination         
KUBE-HOSTPORTS  all  --  anywhere             anywhere             /* kube hostport portals */ ADDRTYPE match dst-t
ype LOCAL
Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
MASQUERADE  all  --  anywhere            !10.0.0.0/8           /* kubenet: SNAT for outbound traffic from cluster *
/ ADDRTYPE match dst-type !LOCAL
MASQUERADE  all  --  loopback/8           anywhere             /* SNAT for localhost access to hostports */
Chain KUBE-HOSTPORTS (2 references)
target     prot opt source               destination         
KUBE-HP-D4IUOLA6Z74GXNRZ  tcp  --  anywhere             anywhere             /* nginx_e2e-tests-runtime-conformance
-mitqy hostport 8001 */ tcp dpt:8001
Chain KUBE-HP-D4IUOLA6Z74GXNRZ (1 references)
target     prot opt source               destination         
KUBE-MARK-MASQ  all  --  10.180.0.2           anywhere             /* nginx_e2e-tests-runtime-conformance-mitqy hos
tport 8001 */
DNAT       tcp  --  anywhere             anywhere             /* nginx_e2e-tests-runtime-conformance-mitqy hostport
 8001 */ tcp to:10.180.0.2:80
Chain KUBE-MARK-MASQ (1 references)
target     prot opt source               destination      
```

**Comment by thockin** [comment]
@dcbw or @matchstick do you or yours have any bandwidth to help triage this?  If hostports are really broken, that's really bad.

@Random-Liu are you saying *all* hostport behavior is broken for you or just in that e2e case?

**Comment by Random-Liu** [comment]
@thockin For now, it only doesn't work in our node e2e test framework.

In the regular e2e test framework, many master components are using hostport, and they are working well. So it looks like that some configuration is needed to support host port, but we've no idea what is missing in the node e2e framework.

**Comment by bprashanth** [comment]
The work here is debugging the diff between e2e and node e2e clusters, I suspect the packets are getting trapped by some rule before hitting the ones in https://github.com/kubernetes/kubernetes/issues/29039#issuecomment-234601574 (because those look alright at first glance). Also I don't think docker deletes iptables rules if restarted with --iptables=false.

**Comment by yujuhong** [comment]
> Also I don't think docker deletes iptables rules if restarted with --iptables=false.

I may be wrong, bu I think docker is installed and started by default on the node. Unless we manually remove the iptable rules created by docker, we'll need to build a new image.

By the way, I tried running kube-proxy on the node, and then the host port magically worked.
Is it possible that host port relies on kube-proxy mucking with the iptable?

I find the comment below in `hostport.go`
https://github.com/kubernetes/kubernetes/blob/ef0c9f0c5b8efbba948a0be2c98d9d2e32e0b68c/pkg/kubelet/network/hostport/hostport.go#L255
```
	// Assuming the node is running kube-proxy in iptables mode
	// Reusing kube-proxy's KubeMarkMasqChain for SNAT
	// TODO: let kubelet manage KubeMarkMasqChain. Other components should just be able to use it
```

**Comment by Random-Liu** [comment]
@yujuhong Cool. Looks like that should be the reason!
Kubenet should not rely on kube-proxy.

**Comment by bprashanth** [comment]
I suppose we're just assuming the chain exists and we should create it if it doesn't

**Comment by thockin** [comment]
The problem with "if it doesn't exist, create it" is that the flag to set which bit is used.  That flag is on the kube-proxy.

So the cleanup here is to move that flag to kubelet and make kubelet be responsible for setting up the rule, and let kube-proxy "just assume it exists".  Can we think of any other traps here?

Can we rally some community help on this?  @marun was asking about other things he could help with...

**Comment by bprashanth** [comment]
Do we actually need the masquerade for anything other than lo traffic? (which I think we need for the return trip otherwise the packet coming out of cbr0 will get handed to eth0, which never saw the "to" packet)

**Comment by thockin** [comment]
Sorry, do you mean in general or just for the node-e2e ?

**Comment by bprashanth** [comment]
In general for host port, do we need masquerade?

**Comment by thockin** [comment]
for host ports, I think we only need masquerade for when a container talks
to its own node on a hostport that loops back to itself, if I recall.

**Comment by bprashanth** [comment]
That's what i meant by "lo traffic". I don't think we even need it if we initally use the node ip, because eth0 won't think it's martian (someone from node team can verify by curling node-ip).

**Comment by thockin** [comment]
There *is* some case where we needed it, or I would never have added it.  I
just can't recall what..

**Comment by bprashanth** [comment]
fyi https://github.com/kubernetes/kubernetes/issues/29742

**Comment by freehan** [comment]
Will move KubeMarkMasqChain management into kubelet.

**Comment by freehan** [comment]
I believe this has been fixed by #30814. Closing this issue.
